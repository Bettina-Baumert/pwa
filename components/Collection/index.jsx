import React from 'react';
import PropTypes from 'prop-types';
import Item from './components/Item';

/**
 * The collection component.
 * @param {Object} props The component props.
 * @param {React.Node} props.children The collections child components.
 * @param {string} props.className The style for the container.
 * @param {string} props.container The tag type of the container.
 * @param {string} props.elementIndexProp The name of the elements index prop.
 * @param {string} props.firstElementProp The indicator prop whether this is the first element.
 * @param {string} props.lastElementProp The indicator prop whether this the last element.
 * @returns {JSX}
 */
const Collection = (props) => {
  const { elementIndexProp, firstElementProp, lastElementProp } = props;
  const elements = React.Children.map(props.children, (element, index) => (
    <Item
      element={element}
      elementIndexProp={elementIndexProp}
      firstElementProp={firstElementProp}
      lastElementProp={lastElementProp}
      elementIndex={index}
      isFirstElement={index === 0}
      isLastElement={index === (React.Children.count(props.children) - 1)}
    />
  ));

  return React.createElement(props.container, { className: props.className }, elements);
};

Collection.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  container: PropTypes.string,
  elementIndexProp: PropTypes.string,
  firstElementProp: PropTypes.string,
  lastElementProp: PropTypes.string,
};

Collection.defaultProps = {
  children: null,
  className: null,
  container: 'div',
  elementIndexProp: null,
  firstElementProp: null,
  lastElementProp: null,
};

Collection.Element = Item;

export default Collection;
