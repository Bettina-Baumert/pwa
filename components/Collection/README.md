# Collection
---

This component is an abstract component intended for any collection of uniform components that should
be rendered (e.g. ordered/unordered lists).


## Getting Started

```
import Collection from 'Components/Collection';

<Collection elementIndexProp="myIndex" firstElementProp="amIFirst" lastElementProp="amILast">
  <MyComponent label="my first component"/>
  <MyComponent label="my second component"/>
  ...
</Collection>
```

The example shown above would render a collection of `MyComponent` elements. It is assumed that `MyComponent`
accepts the properties `myIndex`, `amIFirst` and `amILast` to gain information about their respective position
in the collection.

## Props

### className

_Type_: `string`<br>
_Defaults_: `''`<br>

Styling for the container element.

###### Usage:

```
import { Link } from 'Library/components';
import style from './my-awesome.style';

<Collection className={style}>
  <span>foo</span>
</Collection>
```

### container

_Type_: `string`<br>
_Defaults_: `'div'`<br>

The tag of the element containing the collection elements.

###### Usage:

```
<Collection container="ul">
  <li>I'm a list element!</li>
  <li>Another list element.</li>
</Collection>
```

The expected DOM result of the code shown above would be:

```
<ul>
  <li>I'm a list element!</li>
  <li>Another list element.</li>
</ul>
```

### elementIndexProp

_Type_: `string`<br>
_Defaults_: `null`<br>

The name of the child property that should receive the current index of the element.

###### Usage:

```
import { Link } from 'Library/components';

const MyComponent = props => (
  <span>MyComponent #{props.myIndex}</span>
);

<Collection elementIndexProp="myIndex">
  <MyComponent />
  <MyComponent />
  <MyComponent />
  <MyComponent />
</Collection>
```

The code above would result in the following DOM:

```
<div>
  <span>MyComponent #0</span>
  <span>MyComponent #1</span>
  <span>MyComponent #2</span>
  <span>MyComponent #3</span>
</div>
```

### firstElementProp

_Type_: `string`<br>
_Defaults_: `null`<br>

The name of the child property that should receive information about if this is the first element in
the collection.

###### Usage:

```
import { Link } from 'Library/components';

const MyComponent = props => (
  <span>{props.amIFirst ? `I'm the first element.` : `I'm not the first one.`}</span>
);

<Collection firstElementProp="amIFirst">
  <MyComponent />
  <MyComponent />
  <MyComponent />
  <MyComponent />
</Collection>
```

The code above would result in the following DOM:

```
<div>
  <span>I'm the first element.</span>
  <span>I'm not the first one.</span>
  <span>I'm not the first one.</span>
  <span>I'm not the first one.</span>
</div>
```

### lastElementProp

_Type_: `string`<br>
_Defaults_: `null`<br>

The name of the child property that should receive information about if this is the first element in
the collection.

###### Usage:

```
import { Link } from 'Library/components';

const MyComponent = props => (
  <span>{props.amILast ? "I'm the last element." : "I'm not the last one."}</span>
);

<Collection lastElementProp="amILast">
  <MyComponent />
  <MyComponent />
  <MyComponent />
  <MyComponent />
</Collection>
```

The code above would result in the following DOM:

```
<div>
  <span>I'm not the last one.</span>
  <span>I'm not the last one.</span>
  <span>I'm not the last one.</span>
  <span>I'm the last element!</span>
</div>
```
---
