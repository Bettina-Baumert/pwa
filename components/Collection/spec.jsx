/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { PropTypes } from 'react';
import { shallow, mount } from 'enzyme';
import Collection from './index';

/**
 * Test element component.
 * @param {Object} props The component properties.
 * @returns {JSX}
 */
const TestElement = props => (<li>
  {props.testData}
  Item no. {props.myIndex}
  {props.amIFirst ? (<span className="first">(First element)</span>) : ''}
  {props.amILast ? (<span className="last">(Last element)</span>) : ''}
</li>);

TestElement.propTypes = {
  amIFirst: PropTypes.bool,
  amILast: PropTypes.bool,
  myIndex: PropTypes.number,
  testData: PropTypes.string,
};

TestElement.defaultProps = {
  amIFirst: false,
  amILast: false,
  myIndex: 0,
  testData: '',
};

describe('<Collection />', () => {
  it('renders with children', () => {
    const numChildren = 5;
    const wrapper = mount(
      <Collection elementIndexProp="myIndex" lastElementProp="amILast">
        <TestElement />
        <TestElement />
        <TestElement />
        <TestElement />
        <TestElement />
      </Collection>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(TestElement).length).toBe(numChildren);
  });

  it('renders without children', () => {
    const wrapper = shallow(
      <Collection />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(TestElement).length).toBe(0);
  });

  it('renders the correct container tag', () => {
    const wrapper = shallow(
      <Collection container="ol" />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ol').length).toBe(1);
  });

  it('correctly forwards the element props and merges the additional props', () => {
    const wrapper = mount(
      <Collection elementIndexProp="myIndex">
        <TestElement testData="first" />
        <TestElement testData="second" />
        <TestElement testData="third" />
      </Collection>
    );

    const testElements = wrapper.find(TestElement);

    expect(wrapper).toMatchSnapshot();

    expect(testElements.get(0).props.testData).toBe('first');
    expect(testElements.get(1).props.testData).toBe('second');
    expect(testElements.get(2).props.testData).toBe('third');

    expect(testElements.get(0).props.myIndex).toBe(0);
    expect(testElements.get(1).props.myIndex).toBe(1);
    expect(testElements.get(2).props.myIndex).toBe(2);
  });

  it('receives the first/last props', () => {
    const wrapper = mount(
      <Collection elementIndexProp="myIndex" lastElementProp="amILast" firstElementProp="amIFirst">
        <TestElement testData="first" />
        <TestElement testData="second" />
        <TestElement testData="third" />
      </Collection>
    );

    const testElements = wrapper.find(TestElement);

    expect(wrapper).toMatchSnapshot();

    expect(testElements.get(0).props.amIFirst).toBe(true);
    expect(testElements.get(0).props.amILast).toBe(false);

    expect(testElements.get(1).props.amIFirst).toBe(false);
    expect(testElements.get(1).props.amILast).toBe(false);

    expect(testElements.get(2).props.amIFirst).toBe(false);
    expect(testElements.get(2).props.amILast).toBe(true);
  });
});
