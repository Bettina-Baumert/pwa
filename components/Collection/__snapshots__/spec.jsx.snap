// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collection /> correctly forwards the element props and merges the additional props 1`] = `
<Collection
  className={null}
  container="div"
  elementIndexProp="myIndex"
  firstElementProp={null}
  lastElementProp={null}
>
  <div
    className={null}
  >
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="first"
        />
      }
      elementIndex={0}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={true}
      isLastElement={false}
      key=".0"
      lastElementProp={null}
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={0}
        testData="first"
      >
        <li>
          first
          Item no. 
          0
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="second"
        />
      }
      elementIndex={1}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={false}
      key=".1"
      lastElementProp={null}
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={1}
        testData="second"
      >
        <li>
          second
          Item no. 
          1
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="third"
        />
      }
      elementIndex={2}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={true}
      key=".2"
      lastElementProp={null}
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={2}
        testData="third"
      >
        <li>
          third
          Item no. 
          2
        </li>
      </TestElement>
    </CollectionItem>
  </div>
</Collection>
`;

exports[`<Collection /> receives the first/last props 1`] = `
<Collection
  className={null}
  container="div"
  elementIndexProp="myIndex"
  firstElementProp="amIFirst"
  lastElementProp="amILast"
>
  <div
    className={null}
  >
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="first"
        />
      }
      elementIndex={0}
      elementIndexProp="myIndex"
      firstElementProp="amIFirst"
      isFirstElement={true}
      isLastElement={false}
      key=".0"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={true}
        amILast={false}
        myIndex={0}
        testData="first"
      >
        <li>
          first
          Item no. 
          0
          <span
            className="first"
          >
            (First element)
          </span>
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="second"
        />
      }
      elementIndex={1}
      elementIndexProp="myIndex"
      firstElementProp="amIFirst"
      isFirstElement={false}
      isLastElement={false}
      key=".1"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={1}
        testData="second"
      >
        <li>
          second
          Item no. 
          1
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData="third"
        />
      }
      elementIndex={2}
      elementIndexProp="myIndex"
      firstElementProp="amIFirst"
      isFirstElement={false}
      isLastElement={true}
      key=".2"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={true}
        myIndex={2}
        testData="third"
      >
        <li>
          third
          Item no. 
          2
          <span
            className="last"
          >
            (Last element)
          </span>
        </li>
      </TestElement>
    </CollectionItem>
  </div>
</Collection>
`;

exports[`<Collection /> renders the correct container tag 1`] = `
<ol
  className={null}
/>
`;

exports[`<Collection /> renders with children 1`] = `
<Collection
  className={null}
  container="div"
  elementIndexProp="myIndex"
  firstElementProp={null}
  lastElementProp="amILast"
>
  <div
    className={null}
  >
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData=""
        />
      }
      elementIndex={0}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={true}
      isLastElement={false}
      key=".0"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={0}
        testData=""
      >
        <li>
          Item no. 
          0
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData=""
        />
      }
      elementIndex={1}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={false}
      key=".1"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={1}
        testData=""
      >
        <li>
          Item no. 
          1
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData=""
        />
      }
      elementIndex={2}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={false}
      key=".2"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={2}
        testData=""
      >
        <li>
          Item no. 
          2
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData=""
        />
      }
      elementIndex={3}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={false}
      key=".3"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={false}
        myIndex={3}
        testData=""
      >
        <li>
          Item no. 
          3
        </li>
      </TestElement>
    </CollectionItem>
    <CollectionItem
      element={
        <TestElement
          amIFirst={false}
          amILast={false}
          myIndex={0}
          testData=""
        />
      }
      elementIndex={4}
      elementIndexProp="myIndex"
      firstElementProp={null}
      isFirstElement={false}
      isLastElement={true}
      key=".4"
      lastElementProp="amILast"
    >
      <TestElement
        amIFirst={false}
        amILast={true}
        myIndex={4}
        testData=""
      >
        <li>
          Item no. 
          4
          <span
            className="last"
          >
            (Last element)
          </span>
        </li>
      </TestElement>
    </CollectionItem>
  </div>
</Collection>
`;

exports[`<Collection /> renders without children 1`] = `
<div
  className={null}
/>
`;
